tests/testthat/test-sma-sim.R

```diff
@@ -1,158 +1,76 @@
 library(testthat)
 library(hydromad)
+library(patrick)
 
 context("Soil Moisture Accounting models")
 
+# Set up
 set.seed(0)
 warmup <- 100
 P <- ts(pmax(rnorm(200), 0))
 E <- ts(20 + 10 * cos((1:200) / 20))
 Q <- P * E * runif(P)
 DATA <- cbind(P = P, E = E)
+DATAE <- DATA
+DATAE[, "E"] <- scale(DATAE[, "E"]) * 2
 ## some SMAs require Q also in simulation
 DATAQ <- cbind(P = P, Q = Q)
 
-test_that("cmd simulation is the same in R and C", {
-  set.seed(0)
-  mod0 <- hydromad(DATA, sma = "cmd")
-  for (mod in simulate(mod0, 5)) {
-    Csim <- predict(mod)
-    expect_true(all(Csim >= 0))
-    hydromad.options(pure.R.code = TRUE)
-    pureRsim <- predict(mod)
-    hydromad.options(pure.R.code = FALSE)
-    expect_equal(Csim, pureRsim)
-  }
-})
 
-test_that("cmd power-form simulation is the same in R and C", {
-  set.seed(0)
-  mod0 <- hydromad(DATA, sma = "cmd", shape = 2)
-  for (mod in simulate(mod0, 5)) {
-    Csim <- predict(mod)
-    expect_true(all(Csim >= 0))
-    hydromad.options(pure.R.code = TRUE)
-    pureRsim <- predict(mod)
-    hydromad.options(pure.R.code = FALSE)
-    expect_equal(Csim, pureRsim)
-  }
-})
-
-test_that("cwi simulation is the same in R and C", {
-  set.seed(0)
-  mod0 <- hydromad(DATA, sma = "cwi")
-  for (mod in simulate(mod0, 5)) {
-    Csim <- predict(mod)
-    expect_true(all(Csim >= 0))
-    hydromad.options(pure.R.code = TRUE)
-    pureRsim <- predict(mod)
-    hydromad.options(pure.R.code = FALSE)
-    expect_equal(Csim, pureRsim)
-  }
-})
-
-test_that("bucket simulation is the same in R and C", {
-  set.seed(0)
-  mod0 <- hydromad(DATA, sma = "bucket")
-  for (mod in simulate(mod0, 5)) {
-    Csim <- predict(mod)
-    expect_true(all(Csim >= 0))
-    hydromad.options(pure.R.code = TRUE)
-    pureRsim <- predict(mod)
-    hydromad.options(pure.R.code = FALSE)
-    expect_equal(Csim, pureRsim)
-  }
-})
-
-test_that("gr4j simulation is the same in R and C", {
-  set.seed(0)
-  mod0 <- hydromad(DATA,
-    sma = "gr4j", routing = "gr4jrouting",
-    etmult = 0.1, S_0 = 0.5, R_0 = 0
+with_parameters_test_that(
+  "cmds are the same in R and C",
+  {
+    set.seed(0)
+    n_iter <- 5
+    for (mod in simulate(mod, n_iter)) {
+      if (test_name == "sma = \"gr4j\"") {
```

Comment:
would it be more elegant if the cases passed in a boolean `trim_na`, and we use `exists` to default it to false? We could then just put an if around the predict calls rather than the entire block.

Reply:
I'm not sure why the na.trim is required in the first place - the gr4j implementation should be taking care of making sure the output from R and C implementations are identical. 
So I'll merge this as is for now, and I've added an issue: https://github.com/josephguillaume/hydromad/issues/161